rtmp-services: Add multitrack video configuration support

palana opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
These changes are prerequisites for the upcoming multitrack video/Enhanced Broadcasting support, and add support for eRTMP multitrack video configuration URL (and related disclaimers) to `rtmp-services` (and `libobs`)

### Motivation and Context
Twitch/Amazon IVS uses server side configuration for eRTMP Enhanced Broadcasting/multitrack video stream configuration, so users don't have to manually configure renditions being sent

`ertmp_configuration_url` (and `ertmp_configuration_name`) are exposed in the service json configuration to allow other services to adopt the same configuration mechanism without additional code changes

### How Has This Been Tested?
This is in use in the Twitch Enhanced Broadcasting Beta

### Types of changes
<!--- - Bug fix (non-breaking change which fixes an issue) -->
- New feature (non-breaking change which adds functionality)
<!-- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

tytan652 commented on libobs/obs-service.h:
```diff
@@ -43,6 +43,7 @@ enum obs_service_connect_info {
 	OBS_SERVICE_CONNECT_INFO_PASSWORD = 6,
 	OBS_SERVICE_CONNECT_INFO_ENCRYPT_PASSPHRASE = 8,
 	OBS_SERVICE_CONNECT_INFO_BEARER_TOKEN = 10,
+	OBS_SERVICE_CONNECT_INFO_CONFIG_URL = 12,
```
This does not look like being part of the eRTMP spec but something alike to a custom behavior for a specific service.

`obs_service_connect_info` is meant to try to abstract filed between protocols not service-specific/custom. behavior.

tt2468 replied:
Backing this up. This enum is intended to be used to describe parameters for configuring the protocol being used for streaming. The configuration URL used for Enhanced Broadcast is an out of scope of the RTMP protocol itself, just like  the URL `https://ingest.twitch.tv/ingests` is, so this system should not be used to pass this data around.

palana replied:
moved this functionality in https://github.com/obsproject/obs-studio/pull/10543/commits/06d758a0d5bcc25f36376047da2c6d364ee519ef (<https://github.com/obsproject/obs-studio/pull/10543#event-12615445003>)

RytoEX replied:
This change was reverted, so this feedback was addressed. Resolving.

RytoEX commented on plugins/rtmp-services/data/locale/en-US.ini:
```diff
@@ -8,3 +8,5 @@ UseAuth="Use authentication"
 Username="Username"
 Password="Password"
 ShowAll="Show all services"
+MultitrackVideo.TwitchDisclaimer="Enhanced Broadcasting automatically optimizes your settings to encode and send multiple video qualities to Twitch, powering the best possible viewer experience. Selecting this option will send Twitch information about your computer and software setup. <a href='https://help.twitch.tv/s/article/multiple-encodes'>Learn More</a>"
+MultitrackVideo.Disclaimer="%1 automatically optimizes your settings to encode and send multiple video qualities to %2, powering the best possible viewer experience. Selecting this option will send %2 information about your computer and software setup."
```
I'm wondering if we need two distinct strings here or if we could just parameterize one (already done with `MultitrackVideo.Disclaimer`) and add a second parameterized string for the "Learn More" link that can be optionally configured and appended to the first string before displaying it. That would avoid hard-coding a nearly identical service-specific string here.

palana replied:
changed this in https://github.com/obsproject/obs-studio/pull/10543/commits/72568fcdb031a14deee0fe45a678844d9e9bd26f (<https://github.com/obsproject/obs-studio/pull/10543#event-12615445003>)

RytoEX replied:
Thanks. Resolving.

RytoEX reviewed (commented):
I am curious about the locale string here and if we can remove the service-specific one. No other notes/nits at this time.

WizardCM commented on plugins/rtmp-services/data/schema/service-schema-v5.json:
```diff
@@ -198,6 +198,14 @@
                         "$ref": "#/definitions/saneUrl",
                         "description": "Link that provides additional info about the service, presented in the UI as a button next to the services dropdown."
                     },
+                    "ertmp_configuration_url": {
```
As this is not part of the eRTMP spec, it'd be worth naming this something a little more generic. For example, `custom_encoding_configuration_url`.

Warchamp7 replied:
Is this endpoint meant to be exclusively used for multitrack video? Or would it be reasonable for this to be something like `ingest_configuration_url` or `recommended_configuration_url`

palana replied:
how about `multitrack_video_configuration_url`? this is only required for enhanced broadcasting/multitrack video, though the mechanism doesn't enforce that multiple video tracks are being used or similar

tt2468 replied:
I would like to personally suggest just `track_configuration_url`. I could foresee audio tracks being handled through this URL in the future too.

RytoEX replied:
Since all of the code work has been focused on multitrack video, I am fine with `multitrack_video_configuration_url` unless someone can justify why it should be more generic. If it should/must be more generic, I could understand `track_configuration_url` or `multitrack_configuration_url`.

palana replied:
renamed things to `multitrack_video(_configuration_url)` in https://github.com/obsproject/obs-studio/pull/10543#commits-pushed-ae696c2

RytoEX replied:
Resolving this thread for now. If there are still concerns about the naming, please feel free to comment or open a new feedback thread.

derrod commented:
Also would echo the sentiment that calling it `ertmp` is misleading as its not part of that spec, and is in theory protocol-agnostic, e.g. it could be used with MoQ in the future.

WizardCM commented on plugins/rtmp-services/data/schema/service-schema-v5.json:
```diff
@@ -198,6 +198,14 @@
+                        "$ref": "#/definitions/saneUrl",
+                        "description": "Accessed for multitrack video auto configuration"
+                    },
+                    "ertmp_configuration_name": {
```
Ideally, any user-facing text should be supported by our translation system. As this cannot be, this should not be configurable text, but rather a generic platform/protocol-agnostic field title in our translation file.

RytoEX replied:
Putting this text in our translation system would require jumping through extra hoops, or us having to merge and maintain strings in our codebase. I'm not sure how much extra work that would entail, but I do not consider this feedback a blocker at this time.

murillo128 commented:
Th PR title is misleading and I think some of the comments on the code. 

While the intention may be to add configure multitrack video, this PR and the code in only implements setting a custom configuration URL for an RTMP provider . 

Currently these seems locked to twitch and only configuring multitrack video but, if merged, I think this should be generalized (and "standarized") so it doesn't mention twitch and we allow to configure other stuff different than multitrack video.

RytoEX commented on plugins/rtmp-services/rtmp-common.c:
```diff
@@ -531,6 +536,48 @@ static void update_protocol(json_t *service, obs_data_t *settings)
+					 "ertmp_multitrack_video_name"));
+	dstr_replace(&str, "%2", name);
+	if (learn_more_link.array) {
+		dstr_replace(&str, "%3", learn_more_link.array);
```
Instead of `dstr_replace`, could this not be `dstr_cat` or `dstr_catf` or something? I would think that, if a service had not specified `learn_more_link`, the string would have a lingering `%3` in it on user display.

palana replied:
oh, yes, you're right; at the very least the `%3` needs to be removed when `learn_more_link` is empty

palana replied:
changed in https://github.com/obsproject/obs-studio/pull/10543/commits/9608ddfc056bd2d21bce092ce736961acea0eff0

`dstr_catf` seems tempting, but I'd rather not pass external strings as format arg to `printf`-like functions

RytoEX replied:
Thanks. Resolving.

Lain-B commented:
One of these days we'll have to work on that big service update to make things easier for services to share features as well as to make things easier for individual services to customize (and get that RFC through), but this has my approval

RytoEX commented:
> Th PR title is misleading and I think some of the comments on the code.
> 
> While the intention may be to add configure multitrack video, this PR and the code in only implements setting a custom configuration URL for an RTMP provider .
> 
> Currently these seems locked to twitch and only configuring multitrack video but, if merged, I think this should be generalized (and "standarized") so it doesn't mention twitch and we allow to configure other stuff different than multitrack video.

I've adjusted the PR title.

Twitch is contributing the code, so it aligns with their implementation. That said, there is nothing stopping other platforms/providers from implementing this. See [Twitch's Help Article on Multiple Encodes](https://help.twitch.tv/s/article/multiple-encodes), specifically:
> If you’re a third-party video tool or platform and want help, please [get in touch.](https://docs.google.com/forms/d/e/1FAIpQLSfaYE33ZyIBJDH7B34CNWm-aZkZsdAlR0pqFihHIOWo3fiSfQ/viewform)

None of the actual code mentions Twitch anymore. The only places that do are examples in the JSON schema and the strings in the services.json file in Twitch's entry, where they are appropriate.

@palana Please squash the commits here.

palana commented:
> > Th PR title is misleading and I think some of the comments on the code.
> > While the intention may be to add configure multitrack video, this PR and the code in only implements setting a custom configuration URL for an RTMP provider .
> > Currently these seems locked to twitch and only configuring multitrack video but, if merged, I think this should be generalized (and "standarized") so it doesn't mention twitch and we allow to configure other stuff different than multitrack video.
> 
> I've adjusted the PR title.

thanks

> Twitch is contributing the code, so it aligns with their implementation. That said, there is nothing stopping other platforms/providers from implementing this. See [Twitch's Help Article on Multiple Encodes](https://help.twitch.tv/s/article/multiple-encodes), specifically:
> 
> > If you’re a third-party video tool or platform and want help, please [get in touch.](https://docs.google.com/forms/d/e/1FAIpQLSfaYE33ZyIBJDH7B34CNWm-aZkZsdAlR0pqFihHIOWo3fiSfQ/viewform)
> 
> None of the actual code mentions Twitch anymore. The only places that do are examples in the JSON schema and the strings in the services.json file in Twitch's entry, where they are appropriate.
> 
> @palana Please squash the commits here.

@RytoEX: done (also rebased on master)

RytoEX commented:
@palana Please see @Warchamp7 's [review comment](https://github.com/obsproject/obs-studio/pull/10634#discussion_r1590108405) on #10634.

palana commented on plugins/rtmp-services/data/locale/en-US.ini:
```diff
@@ -8,3 +8,5 @@ UseAuth="Use authentication"
 Username="Username"
 Password="Password"
 ShowAll="Show all services"
+MultitrackVideo.Disclaimer="%1 automatically optimizes your settings to encode and send multiple video qualities to %2, powering the best possible viewer experience. Selecting this option will send %2 information about your computer and software setup."
```
re https://github.com/obsproject/obs-studio/pull/10634#pullrequestreview-2039534116
@Warchamp7, @RytoEX: so should we change the string to this
```suggestion
MultitrackVideo.Disclaimer="%1 automatically optimizes your settings to encode and send multiple video qualities to %2. Selecting this option will send %2 information about your computer and software setup."
```
?

RytoEX replied:
Yes.

RytoEX reviewed (changes requested):
Commit message nit:
>rtmp-services: Allow loading Multitrack Video configuration from services.json

This commit message subject is too long.

Instead, use this:
```
rtmp-services: Allow loading Multitrack Video config from JSON

Add support for loading Multitrack Video configuration data from
services.json.
```

palana commented:
> Commit message nit:
> 
> > rtmp-services: Allow loading Multitrack Video configuration from services.json
> 
> This commit message subject is too long.
> 
> Instead, use this:
> 
> ```
> rtmp-services: Allow loading Multitrack Video config from JSON
> 
> Add support for loading Multitrack Video configuration data from
> services.json.
> ```

@RytoEX: done

RytoEX reviewed (approved):
Looks good to me.
